engineering/architecture/numbers.md

```diff
@@ -0,0 +1,106 @@
+---
+title: Number display and entry
+description: Goals and strategies for the display and entry of numerical data
+published: true
+date: 2022-03-15T00:00:00.000Z
+tags: 
+editor: markdown
+dateCreated: 2022-03-15T00:00:00.000Z
+---
+
+## Product goals
+
+- **(Goal A)** Mathesar is capable of displaying and accepting entry for all [numbers which Postgres is capable of storing](https://www.postgresql.org/docs/current/datatype-numeric.html).
+
+- **(Goal B)** Aside from the numbers, `NaN`, `-Infinity`, and `Infinity`, users are able to type all numbers into Mathesar exactly as Mathesar displays those numbers to them.
+
+- **(Goal C)** When a user enters data which cannot be accurately stored in the column exactly as-entered, then Mathesar stores the nearest acceptable value and clearly communicates to stored value to the user.
+
+    For example, a `float8` column cannot accurately store the value `9999999999999999` (because it will instead be stored as `1e+16`). In this example, Mathesar will store `1e+16`, displaying that value back to the user after entry.
+
+- **(Goal D)** Mathesar displays numbers and accepts entry of numbers in the user's desired locale, including the correct decimal separator and grouping separators.
+
+- **(Goal E)** Mathesar auto-formats numbers as the user types.
+
+- **(Goal F)** Mathesar accepts entry of numbers in scientific notation and displays numbers in scientific notation when the user is likely to prefer it. For example `1E100` instead of 1 followed by 100 zeros.
+
+## Implementation strategies
+
+### API
+
+- The API transmits numbers as strings because most JSON parsers are not capable of accurately parsing high precision numbers or special Postgres numbers stored plainly in JSON. The format for such strings is described below.
+
+### Canonical stringified number format
+
+Throughout our stack, we may want to represent numbers as strings. When doing so, we follow a format very similar to numbers in JSON.
+
+- The decimal separator is an ASCII dot.
+- The minus sign is an ASCII hyphen.
+- The exponent separator is an uppercase "E".
+- Grouping separators are disallowed.
+- The strings `"NaN"`, `"-Infinity"`, and `"Infinity"` are used to represent special Postgres numbers. Note that this is one way in which our stringified number format differs from JSON -- raw values like `NaN` are _not_ valid inside JSON.
+- The string `"NULL"` should _not_ be used. In JSON, the _value_ `null` should be used instead.
+- Examples:
+
+    - `"NaN"`
+    - `"-Infinity"`
+    - `"Infinity"`
+    - `"0"`
+    - `"1234.5"`
+    - `"99999999999999999999999999999999"`
+    - `"-1.2E-15"`
+
+
+### Locale
+
+- To satisfy Goal D, we infer the user's desired locale from their _browser_. The back-end is not aware of the user's locale unless the front end passes data about its in the API request.
```

Comment:
I assume you'll update this to include display options.

Reply:
Fixed in 86b2315ec28d088372c8c0e42bc492f8745b2cd8